### DISCUSSION TITLE
Use Mock Service Worker (msw) to mock network in component (and e2e) testing?

### DISCUSSION THREAD

**Discussion by bhootd** [webdriverio/webdriverio#10170] [Q&A] [answered]
From what I understand, msw should be initialised in one of the hooks.

But msw also needs that a '/mockServiceWorker.js' file should be served during the run. I can't figure out how to ask wdio to do that.

Are there any examples out there?

**Comment by christian-bromann** [comment]
> Are there any examples out there?

Can you please describe in length what you are trying to do? What are you referring to when you say

> But msw also needs that a '/mockServiceWorker.js' file should be served during the run.

?

**Reply by bhootd** [reply]
msw (https://github.com/mswjs/msw) uses a service worker to intercept and mock network requests. It lets me define a handler per API endpoint, all at one place.

As it uses a service worker, its very much ideal for in-browser testing.

There are two components to make msw work:

1. Code that starts the worker. This would probably go into a wdio hook.
2. A mockServiceWorker.js file which lives along side the code being served. So if the code being tested is being served at `localhost:8988`, then `localhost:8988/mockServiceWorker.js` needs to exist. In case of a webpack bundle, I would write the bundled output to `build` dir, and also copy `mockServiceWorker.js` to the `build` dir. Not sure how to handle this wrt wdio.

**Reply by bhootd** [reply]
I've found `@wdio/static-server-service`. However, it does not seem to use the same port as the runner (which is essential for the file to be served at /mockServiceWorker.js) Is there a way to force `@wdio/static-server-service` to do so?

**Reply by christian-bromann** [reply]
Are you writing e2e or UI/component tests?

**Comment by bhootd** [comment]
I am writing a UI/Component test.

But I would like the mock setup to work for e2e too. That way, I can share that code.

**Reply by christian-bromann** [reply]
These are two different test types. For UI/component tests I recommend to either mock the `fetch` method and for e2e tests use the [mock feature](https://webdriver.io/docs/api/browser/mock). You can theoretically use the mock feature for both.

**Reply by bhootd** [reply]
> You can theoretically use the mock feature for both

I tried that out unsuccessfully before reaching for msw.

I want to share the mock responses between component and e2e tests. That's what has attracted me to WebdriverIO - not having to maintain two test stacks - non-e2e and e2e.

Should I create a separate issue about the failed attempt to use `browser.mock` for component tests?

Do your suggestions also mean that what I want for msw is not possible with WebdriverIO?

**Reply by christian-bromann** [reply]
> I tried that out unsuccessfully

Why was it unsuccessful? Do you have a reproducible example?

> Should I create a separate issue about the failed attempt to use `browser.mock` for component tests?

Let's talk about that here and only create an issue if we know it is a bug.

> I want to share the mock responses between component and e2e tests.

You can totally do that with WebdriverIO.

**Reply by bhootd** [reply]
> Why was it unsuccessful? Do you have a reproducible example?

I ran into the following error:

```
$ DEBUG=true npx wdio run ./wdio.conf.js
2023-04-12T04:40:53.232Z WARN @wdio/config:ConfigParser: pattern ./tests-wdio/**/*.test.js did not match any file

Execution of 1 workers started at 2023-04-12T04:40:53.234Z

2023-04-12T04:40:53.595Z WARN @wdio/config:ConfigParser: pattern ./tests-wdio/**/*.test.js did not match any file
[0-0] RUNNING in chrome - file:///tests-wdio/L2/login.test.jsx
[0-0] %cDownload the React DevTools for a better development experience: https://reactjs.org/link/react-devtools font-weight:bold
[0-0] setting up mocks
[0-0] TypeError in "React Component Tests.should test component with Testing Library"
TypeError: Cannot read properties of undefined (reading 'selenoid:options')
    at Proxy.getPuppeteer (http://localhost:52838/node_modules/webdriverio/build/commands/browser/getPuppeteer.js?v=b32bc9c5:67:52)
    at Proxy.wrapCommandFn (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:72:38)
    at Proxy.<anonymous> (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:210:24)
    at Proxy.mock (http://localhost:52838/node_modules/webdriverio/build/commands/browser/mock.js?v=b32bc9c5:114:20)
    at Proxy.wrapCommandFn (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:72:38)
    at Proxy.mock (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:210:24)
    at Context.<anonymous> (http://localhost:52838/Users/jayesh/project/tests-wdio/L2/login.test.jsx:17:42)
[0-0] FAILED in chrome - file:///tests-wdio/L2/login.test.jsx

 "spec" Reporter:
------------------------------------------------------------------
[chrome 112.0.5615.49 mac os x #0-0] Running: chrome (v112.0.5615.49) on mac os x
[chrome 112.0.5615.49 mac os x #0-0] Session ID: 225900d8885613777f9a9354c795b1ce
[chrome 112.0.5615.49 mac os x #0-0]
[chrome 112.0.5615.49 mac os x #0-0] » file:///tests-wdio/L2/login.test.jsx
[chrome 112.0.5615.49 mac os x #0-0] React Component Tests
[chrome 112.0.5615.49 mac os x #0-0]    ✖ should test component with Testing Library
[chrome 112.0.5615.49 mac os x #0-0]
[chrome 112.0.5615.49 mac os x #0-0] 1 failing (2s)
[chrome 112.0.5615.49 mac os x #0-0]
[chrome 112.0.5615.49 mac os x #0-0] 1) React Component Tests should test component with Testing Library
[chrome 112.0.5615.49 mac os x #0-0] Cannot read properties of undefined (reading 'selenoid:options')
[chrome 112.0.5615.49 mac os x #0-0] TypeError: Cannot read properties of undefined (reading 'selenoid:options')
[chrome 112.0.5615.49 mac os x #0-0]     at Proxy.getPuppeteer (http://localhost:52838/node_modules/webdriverio/build/commands/browser/getPuppeteer.js?v=b32bc9c5:67:52)
[chrome 112.0.5615.49 mac os x #0-0]     at Proxy.wrapCommandFn (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:72:38)
[chrome 112.0.5615.49 mac os x #0-0]     at Proxy.<anonymous> (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:210:24)
[chrome 112.0.5615.49 mac os x #0-0]     at Proxy.mock (http://localhost:52838/node_modules/webdriverio/build/commands/browser/mock.js?v=b32bc9c5:114:20)
[chrome 112.0.5615.49 mac os x #0-0]     at Proxy.wrapCommandFn (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:72:38)
[chrome 112.0.5615.49 mac os x #0-0]     at Proxy.mock (http://localhost:52838/node_modules/@wdio/utils/build/shim.js?v=b32bc9c5:210:24)
[chrome 112.0.5615.49 mac os x #0-0]     at Context.<anonymous> (http://localhost:52838/Users/jayesh/project/tests-wdio/L2/login.test.jsx:17:42)


Spec Files:	 0 passed, 1 failed, 1 total (100% completed) in 00:00:04
```

Looks similar to https://github.com/webdriverio/webdriverio/issues/7780, except that I am not using the watch mode.

I was setting up the `browser.mock` and `mock.respond` in mocha's `before` function inside a `describe` suite.

I could try to reproduce it if you want more details.

**Reply by bhootd** [reply, marked as answer]
While I was digging in #7780, I realised that wdio uses vite dev server to serve and run component tests, and that serving `/mockServiceWorker.js` is not wdio's job, but the job of the underlying vite dev server.

So I put the above file in `public` dir, and msw works now, with mocking and everything!

**Reply by bhootd** [reply]
The only question is what is the right place to activate msw?

Right now, I activate it inside mocha's `describe() -> before()` fixture, but that is a lot of repetition.

msw could be activated before any test runs. My hunch is one of the hooks in `wdio.conf.js`. However, when I put the activation code in one of those hooks, msw complains that `[MSW] Failed to execute `setupWorker` in a non-browser environment. . Consider using `setupServer` for Node.js environment instead.`, implying that the hooks in wdio.conf.js run in non-browser environment.

Does any of the hooks run inside the browser, before any of the tests run?

**Reply by christian-bromann** [reply]
> Does any of the hooks run inside the browser, before any of the tests run?

No, can you use the beforeAll hook from e.g. Mocha?

Otherwise, mind raising an issue requesting to inject HTML, CSS and JS fixtures into the sandbox page when running browser tests?

**Reply by bhootd** [reply]
> No, can you use the beforeAll hook from e.g. Mocha?

Mocha provides [global fixtures](https://mochajs.org/#global-fixtures) as well as [root hooks](https://mochajs.org/#root-hook-plugins). Both can be loaded via mocha config's `require` option.

wdio's runners also theoretically [supports](https://webdriver.io/docs/frameworks#require) the `require` config option as `mochaOpts.require`.

However, while `mochaOpts.require` works in case of the local runner, the browser runner fails with the error:

```
[0-5]  Error:  Test failed due to following error(s):
  - mocha.js: Uncaught TypeError: self[opt] is not a function
    at http://localhost:57323/node_modules/mocha/browser-entry.js:161:18
    at Array.forEach (<anonymous>)
    at Mocha.mocha.setup (http://localhost:57323/node_modules/mocha/browser-entry.js:159:6)
    at new MochaFramework (http://localhost:57323/node_modules/@wdio/browser-runner/build/browser/frameworks/mocha.js?v=b6ac1b03:37:15)
    at http://localhost:57323/node_modules/@wdio/browser-runner/build/browser/frameworks/mocha.js?v=b6ac1b03:236:16
```

`self[opt]` here implies an unsupported config key. Similar error can be raised by using a mistyped key like `timeot` instead of `timeout`.

This rules out using both global fixtures and root hooks of mocha with the browser runner.

The next 'global' level in mocha is the suite-level hook system - `before` and `after`. Activating msw in `before` works. It also makes more sense given that not all component tests need network access to be mocked in the first place. So, those which need network access to be mocked could be grouped into a suite with a msw-activating `before` hook.

---
As an aside, while the local runner supports `mochaOpts.require: ['tests/hooks.js']`, it entirely ignore a string value like `mochaOpts.require: 'tests/hooks.js'`. The [wdio docs say](https://webdriver.io/docs/frameworks#require) `require` supports both `string` and `array`. The browser runner doesn't support `require` key at all.

---

Please note that I am not at all complaining here. I am just documenting my findings for posterity. As is, integrating wdio with mocha is already a complex task. Ensuring that the two runners work the same adds complexity on top of it.

**Reply by bhootd** [reply]
> Otherwise, mind raising an issue requesting to inject HTML, CSS and JS fixtures into the sandbox page when running browser tests?

As I mentioned above, activating msw in the suite-level `before` hook works. That might suffice for now.

While injection could put wdio notches above the other frameworks. in my example, just fixing the mocha hook integration could do the job. The vite dev server used by the browser runner to serve the index.html already does half the job by serving the /mockServiceWorker.js file. The other half is actually activating msw, which is possible only in either the app code, or a pre-tests hook, or as an injected piece of code which also has all the required context.

When I get time, I would love to fix the mocha intgration so that the local and browser runners at on a par with each other w.r.t mocha support.

**Reply by christian-bromann** [reply]
> he browser runner doesn't support `require` key at all.

Can we create an issue for this? We should ensure the experience between local and browser runner are the same.

**Reply by bhootd** [reply]
> Can we create an issue for this?

[Done](https://github.com/webdriverio/webdriverio/issues/10220)

There isn't anything left to discuss here.

I am marking my own comment as the answer because its the closest thing to an answer to the question of this discussion.

Thank you for such a prompt support. Love your work!
